server/src/main/java/org/elasticsearch/index/shard/IndexWriteLoad.java

```diff
@@ -0,0 +1,217 @@
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.ToXContentFragment;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+
+public class IndexWriteLoad implements Writeable, ToXContentFragment {
+    public static final ParseField SHARDS_WRITE_LOAD_FIELD = new ParseField("shards_write_load");
+    public static final ParseField SHARDS_UPTIME_IN_MILLIS = new ParseField("shards_uptime_in_millis");
+    private static final Double UNKNOWN_LOAD = -1.0;
+    private static final long UNKNOWN_UPTIME = -1;
+
+    @SuppressWarnings("unchecked")
+    private static final ConstructingObjectParser<IndexWriteLoad, Void> PARSER = new ConstructingObjectParser<>(
+        "index_write_load_parser",
+        false,
+        (args, unused) -> IndexWriteLoad.create((List<Double>) args[0], (List<Long>) args[1])
+    );
+
+    static {
+        PARSER.declareDoubleArray(ConstructingObjectParser.constructorArg(), SHARDS_WRITE_LOAD_FIELD);
+        PARSER.declareLongArray(ConstructingObjectParser.constructorArg(), SHARDS_UPTIME_IN_MILLIS);
+    }
+
+    public static IndexWriteLoad create(List<Double> shardsWriteLoad, List<Long> shardsUptimeInMillis) {
+        if (shardsWriteLoad.size() != shardsUptimeInMillis.size()) {
+            throw new IllegalArgumentException(
+                "The same number of shards write load and shard uptime should be provided, but "
+                    + shardsWriteLoad
+                    + " "
+                    + shardsUptimeInMillis
+                    + " were provided"
+            );
+        }
+
+        if (shardsWriteLoad.isEmpty()) {
+            throw new IllegalArgumentException("At least one shard write load and uptime should be provided, but none was provided");
+        }
+
+        return new IndexWriteLoad(
+            shardsWriteLoad.stream().mapToDouble(shardLoad -> shardLoad).toArray(),
+            shardsUptimeInMillis.stream().mapToLong(shardUptime -> shardUptime).toArray()
+        );
+    }
+
+    @Nullable
+    public static IndexWriteLoad fromStats(IndexMetadata indexMetadata, @Nullable IndicesStatsResponse indicesStatsResponse) {
+        if (indicesStatsResponse == null) {
+            return null;
+        }
+
+        final IndexStats indexStats = indicesStatsResponse.getIndex(indexMetadata.getIndex().getName());
+        if (indexStats == null) {
+            return null;
+        }
+
+        final int numberOfShards = indexMetadata.getNumberOfShards();
+        final var indexWriteLoadBuilder = IndexWriteLoad.builder(numberOfShards);
+        final var indexShards = indexStats.getIndexShards();
+        for (IndexShardStats indexShardsStats : indexShards.values()) {
+            final var shardStats = Arrays.stream(indexShardsStats.getShards())
+                .filter(stats -> stats.getShardRouting().primary())
```

Comment:
Looks like `indexShardsStats.getPrimary()` could be used instead? But this makes me wonder if we should not use `indexingShardsStats.getTotal()` in order to consider both primary and replica work? In particular if one has been up longer than the other etc.

Reply:
I think that approach could be problematic in cases were we fail to fetch some of the shard stats, it could potentially under-represent the write load of those. I would prefer to either just fetch the primary write-load and try to gather it during read time if it isn't present as you suggested or have a more sophisticated fallback mechanism (i.e. using the `max(shard.write_load)` or `max(shard.active_time)`, wdyt?

Reply:
> I think that approach could be problematic in cases where we fail to fetch some of the shard stats, it could potentially under-represent the write load of those. 

I am not sure that is the case, since the up-time would be shorter too. We would still be averaging the combined write-load over the total up-time duration.

But it is not entirely without compromises in that it could hide a very short uptime (which we may want to disregard). So will resolve this comment for now.